src/LinAlg/hiopLinSolverSparseCUSOLVER.hpp

```diff
@@ -0,0 +1,279 @@
+#include "hiopMatrixSparseTriplet.hpp"
+#include <unordered_map>
+
+#include "hiop_cusolver_defs.hpp"
+#include "klu.h"
+/** implements the linear solver class using nvidia_ cuSolver (GLU refactorization)
+ *
+ i* @ingroup LinearSolvers
+ */
+#if defined(HIOP_USE_RAJA)
+//#include "hiopRajaUmpireUtils.hpp"
+#endif
+
+namespace hiop
+{
+  class hiopLinSolverIndefSparseCUSOLVER: public hiopLinSolverSymSparse
+  {
+    public:
+      //constructor
+      hiopLinSolverIndefSparseCUSOLVER(const int& n, const int& nnz, hiopNlpFormulation* nlp);
+      virtual ~hiopLinSolverIndefSparseCUSOLVER();
+
+      /** Triggers a refactorization of the matrix, if necessary.
+       * Overload from base class. 
+       * In this case, KLU (SuiteSparse) is used to refactor*/
+      int matrixChanged();
+
+      /** solves a linear system.
+       * param 'x' is on entry the right hand side(s) of the system to be solved. On
+       * exit is contains the solution(s).  */
+      bool solve ( hiopVector& x_ );
+
+      /** Multiple rhs not supported yet */
+      virtual bool solve(hiopMatrix& x) 
+      { 
+        assert(false && "not yet supported");
+        return false;
+      }
+
+      int newKLUfactorization();
+    private:
+
+      int      m_;// number of rows of the whole matrix
+      int      n_;// number of cols of the whole matrix
+      int      nnz_;// number of nonzeros in the matrix
+
+      int* kRowPtr_;// row pointer for nonzeros
+      int* jCol_;// column indexes for nonzeros
+      double* kVal_;// storage for sparse matrix
+
+      int* index_covert_CSR2Triplet_;
+      int* index_covert_extra_Diag2CSR_;
+
+      int nFakeNegEigs_;
+      /** needed for cuSolver **/
+
+      cusolverStatus_t sp_status_;
+      cusparseHandle_t handle_ = 0;
+      cusolverSpHandle_t handle_cusolver_ = NULL;
+      cublasHandle_t handle_cublas_;
+
+      cusparseMatDescr_t descr_A_, descr_M_;
+      csrluInfoHost_t info_lu_ = NULL;
+      csrgluInfo_t info_M_ = NULL;
+
+      size_t buffer_size_;
+      size_t size_M_;
+      double* d_work_;
+      int ite_refine_succ_ = 0;
+      double r_nrminf_; //, x_nrminf, b_nrminf;
+
+
+      // KLU stuff
+      int klu_status_;
+      klu_common Common_;
+      klu_symbolic* Symbolic_ = NULL;
+      klu_numeric* Numeric_ = NULL;
+      /*pieces of M */
+      int* mia_ = NULL; 
+      int*  mja_ = NULL;
+
+      /* for GPU data */
+      double* da_;
+      int* dia_;
+      int* dja_;
+      double* devx_;
+      double* devr_;
+      double* drhs_;
+      /* private function: creates a cuSolver data structure from KLU data structures. */
+
+      int createM(const int n, 
+          const int nnzL, 
+          const int* Lp, 
+          const int* Li,
+          const int nnzU, 
+          const int* Up, 
+          const int* Ui);
+      template <typename T>
+          void hiopCheckCudaError(T result,
+          char const *const func,
+          const char *const file,
+          int const line);
```

Comment:
Should probably also be in a seperate header so we can reuse it.

```suggestion
      template <typename T>
      void hiopCheckCudaError(
          T result,
          char const *const func,
          const char *const file,
          int const line);
```